File: services/admin-service/internal/handlers/license_cap.go
package handlers

// GET /admin/license/cap — the MSP soft cap on tenants, for the admin
// console's banner (edition-licensing spec §3, PR 3).
//
// A Core route on purpose: the admin console shell asks on every build, and a
// Core or Enterprise install simply answers "uncapped". The cap itself is
// decided in shared/entitlements (AdmitTenantCreation, on every tenant-creation
// path); this only reports it. Read-only: no advisory lock and no write. The
// grace clock is started on writes (a creation, the licence being recorded, a
// tenant's is_operator changing), never by a poll.

import (
	"database/sql"
	"log"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/vistasecurity/vistaplatform/shared/entitlements"
)

// LicenseCapResponse is the wire shape of GET /admin/license/cap.
type LicenseCapResponse struct {
	// Edition is the active licence's edition: core (no active licence),
	// enterprise or msp.
	Edition string `json:"edition"`
	// Licensed is the licence's max_tenants; null when uncapped.
	Licensed *int `json:"licensed"`
	// Current counts live tenants that are not the operator's own.
	Current int `json:"current"`
	// Operator counts live tenants marked as the operator's own (never counted).
	Operator       int        `json:"operator"`
	GraceStartedAt *time.Time `json:"grace_started_at"`
	// GraceDays is the licence's grace period; null when uncapped.
	GraceDays   *int       `json:"grace_days"`
	GraceEndsAt *time.Time `json:"grace_ends_at"`
	// State is under | grace | blocked | uncapped.
	State string `json:"state"`
}

// NewLicenseCapResponse converts the shared status to the wire shape.
func NewLicenseCapResponse(st entitlements.TenantCapStatus) LicenseCapResponse {
	out := LicenseCapResponse{
		Edition:        string(st.Edition),
		Licensed:       st.Licensed,
		Current:        st.Current,
		Operator:       st.Operator,
		GraceStartedAt: st.GraceStartedAt,
		GraceEndsAt:    st.GraceEndsAt,
		State:          string(st.State),
	}
	if out.Edition == "" {
		out.Edition = string(entitlements.EditionCore)
	}
	if out.State == "" {
		out.State = string(entitlements.TenantCapUncapped)
	}
	if st.Licensed != nil {
		days := st.GraceDays
		out.GraceDays = &days
	}
	return out
}

// GetLicenseCap serves GET /admin/license/cap. It only reads (tenants,
// platform_license, license_cap_grace), in a read-only transaction.
func GetLicenseCap(db *sql.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		st, err := entitlements.ReadTenantCapStatus(c.Request.Context(), db, time.Now())
		if err != nil {
			log.Printf("[license-cap] read: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Could not read the licensed tenant limit"})
			return
		}
		c.JSON(http.StatusOK, NewLicenseCapResponse(st))
	}
}

File: admin-ui-v2/src/lib/license-cap.ts
// The MSP soft cap on tenants — GET /api/v1/admin-service/admin/license/cap
// (edition-licensing spec §3, PR 3).
//
// A Core route every build mounts: Core and Enterprise answer
// `state: "uncapped"`, and only an MSP licence carrying `max_tenants` caps
// anything. The shell's banner (app/license-cap-banner.tsx) and the tenant
// drawer's operator-tenant control both read it through this one query.
//
// The route needs platform.settings, so the query is only issued for an
// operator who holds it; everyone else simply never sees the banner.
import { useQuery } from '@tanstack/react-query';
import type { adminServiceComponents } from '@vistasecurity/api-contract';
import { PLATFORM_PERMISSIONS, usePlatformPermissions } from '@vistasecurity/primitives/platform-auth';
import { clients } from './clients';

export type LicenseCap = adminServiceComponents['schemas']['LicenseCapResponse'];

export const licenseCapKey = ['platform', 'license-cap'] as const;

/**
 * Query options for the licensed-tenant read. Exported apart from the hook so
 * a test can run them through a real QueryClient (same shape as
 * platformEditionQuery in ./edition).
 */
export function licenseCapQuery() {
  return {
    queryKey: licenseCapKey,
    // The grace clock moves in days; a signup elsewhere can start it at any
    // time, so poll gently rather than only on navigation.
    staleTime: 60 * 1000,
    refetchInterval: 5 * 60 * 1000,
    retry: 1,
    queryFn: async (): Promise<LicenseCap> => {
      const { data, error } = await clients.admin.GET('/admin/license/cap', {});
      if (error || !data) throw new Error('Could not read the licensed tenant limit');
      return data;
    },
  };
}

export function useLicenseCap() {
  const { hasPermission } = usePlatformPermissions();
  return useQuery({
    ...licenseCapQuery(),
    enabled: hasPermission(PLATFORM_PERMISSIONS.platform.settings),
  });
}

const DAY_MS = 24 * 60 * 60 * 1000;

/** Whole days of grace left, rounded up, never negative. */
export function graceDaysLeft(graceEndsAt: string | null | undefined, now: Date = new Date()): number {
  if (!graceEndsAt) return 0;
  const ms = new Date(graceEndsAt).getTime() - now.getTime();
  return ms <= 0 ? 0 : Math.ceil(ms / DAY_MS);
}

export type CapBanner = { tone: 'warn' | 'danger'; title: string; detail: string };

/**
 * What the shell banner says, or null when it should not render. Only the two
 * over-the-licence states are shown: under the licence, uncapped, or unknown
 * (no data yet, an error, no permission) all render nothing.
 */
export function capBanner(cap: LicenseCap | undefined, now: Date = new Date()): CapBanner | null {
  if (!cap || cap.licensed == null) return null;
  const counts = `${cap.current} of ${cap.licensed} licensed tenants`;
  if (cap.state === 'grace') {
    const days = graceDaysLeft(cap.grace_ends_at, now);
    return {
      tone: 'warn',
      title: `${counts} — ${days} ${days === 1 ? 'day' : 'days'} of grace left`,
      detail: 'Existing tenants are unaffected. When the grace period ends, new tenants will be refused until the licence is extended or tenants are removed.',
    };
  }
  if (cap.state === 'blocked') {
    return {
      tone: 'danger',
      title: `${counts} — grace period ended, new tenants are blocked`,
      detail: 'Existing tenants keep working. Contact Vista Security to extend your licence, or remove tenants to get back under it.',
    };
  }
  return null;
}
